mino/minogrpc/server.go

```diff
@@ -208,23 +280,169 @@ func (o overlayServer) Stream(stream Overlay_StreamServer) error {
 	return nil
 }
 
+// dispatchMessage creates multiple envelopes based on the relays needed for
+// each recipient. It also notifies the receiver in case the recipient is
+// ourself. This function doesn't send any message, it only returns a map that
+// will allow us to then send the envelopes and create the necessary relays if
+// needed.
+func dispatchMessage(envelope Envelope, sender sender) (map[mino.Address]*Envelope, error) {
+	out := map[mino.Address]*Envelope{}
+
+	tos := make([]mino.Address, len(envelope.To))
+	for i, addrBuf := range envelope.To {
+		tos[i] = sender.receiver.addressFactory.FromText(addrBuf)
+	}
+
+	for _, tobuf := range envelope.GetTo() {
+		to := sender.receiver.addressFactory.FromText(tobuf)
+
+		if to.Equal(sender.me) {
+			sender.receiver.appendMessage(envelope.GetMessage())
+			continue
+		}
+
+		packet := sender.router.MakePacket(sender.me, to, envelope.GetMessage().GetPayload())
+		relay, err := sender.router.Forward(packet)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to find route from %s to %s: %v",
+				sender.me, to, err)
+		}
+
+		env, found := out[relay]
+		if !found {
+			out[relay] = &Envelope{
+				To:      [][]byte{tobuf},
+				Message: envelope.GetMessage(),
+			}
+		} else {
+			env.To = append(env.To, tobuf)
+		}
+	}
+
+	return out, nil
+}
+
+// sendToRelays set up the relays if needed and send the envelopes accordingly.
+func sendToRelays(relayCtx context.Context, sender sender, out map[mino.Address]*Envelope) error {
+	for relayAddr, env := range out {
+
+		itf, ok := sender.relays.Load(relayAddr.String())
+		var relay relayer
+
+		if ok {
+			relay = itf.(relayer)
+		} else {
+			conn, err := sender.connFactory.FromAddress(relayAddr)
```

Comment:
Reminder to check if a new connection is created every time, or if we reuse them.